File: filterPortsDuplicate.php
<?php

// remove duplicate IP from database

require __DIR__ . '/func-proxy.php';

global $isCli;

use PhpProxyHunter\ProxyDB;
use PhpProxyHunter\Scheduler;

if (!$isCli) {
  exit('only CLI allowed');
}

$max_checks = 500; // max proxies to be checked
$maxExecutionTime = 120; // max 120s execution time
$endless = false;
$perform_delete = false;

if ($isCli) {
  $short_opts = "p:m::";
  $long_opts = [
    "proxy:",
    "max::",
    "userId::",
    "lockFile::",
    "runner::",
    "admin::",
    "endless::",
    "delete::"
  ];
  $options = getopt($short_opts, $long_opts);
  $perform_delete = !empty($options['delete']); // --delete=true
  if (!empty($options['max'])) { // --max=100
    $max = intval($options['max']);
    if ($max > 0) {
      $max_checks = $max;
    }
  }
  if (!empty($options['admin']) && $options['admin'] !== 'false') { // --admin=true
    $isAdmin = true;
  }
  if (!empty($options['endless'])) { // --endless=true
    $endless = true;
  }
}

if ($isAdmin || $endless) {
  // set time limit 10 minutes for admin and 600 minutes for endless mode
  $maxExecutionTime = $endless ? 600 * 60 :  10 * 60;
  // disable execution limit
  set_time_limit(0);
}

$lockFilePath = tmp() . "/runners/" . basename(__FILE__, '.php') . ".lock";
if ($endless) {
  $lockFilePath = tmp() . "/" . basename(__FILE__, '.php') . "-endless.lock";
}

$statusFile = __DIR__ . "/status.txt";

// Check if the lock file exists
if (file_exists($lockFilePath)) {
  if ($endless || !$isAdmin) {
    // Exit with a message if another process is running and the conditions are met
    exit(date(DATE_RFC3339) . ' another process still running ' . basename(__FILE__, '.php') . PHP_EOL);
  }
} else {
  // Create the lock file and write the current date
  write_file($lockFilePath, date(DATE_RFC3339));
  // Create or update the status file with the message 'filter-ports'
  write_file($statusFile, 'filter-ports');
}

Scheduler::register(function () use ($lockFilePath, $statusFile) {
  if (file_exists($lockFilePath)) {
    unlink($lockFilePath);
  }
  file_put_contents($statusFile, 'idle');
}, "z_Exit_" . md5(__FILE__));

$db = new ProxyDB();
$pdo = $db->db->pdo;

// Step 1: Identify and process duplicates based on IP address in batches
$batchSize = 1000; // Adjust batch size as needed
$start = 0;
$duplicateIds = [];

do {
  // Fetch a batch of duplicate proxies
  $stmt = $pdo->prepare("SELECT ip, COUNT(*) AS count_duplicates
  FROM (
    SELECT SUBSTR(proxy, 0, INSTR(proxy, ':')) AS ip
    FROM proxies
    WHERE status != 'active'
    AND status != 'untested'
    AND last_check < datetime('now', '-7 days')
  ) AS filtered_proxies
  GROUP BY ip
  HAVING COUNT(*) > 1
  ORDER BY RANDOM()
  LIMIT :start, :batchSize");
  $stmt->bindParam(':start', $start, PDO::PARAM_INT);
  $stmt->bindParam(':batchSize', $batchSize, PDO::PARAM_INT);
  $stmt->execute();
  $duplicateIpCounts = $stmt->fetchAll(PDO::FETCH_ASSOC);
  $startTime = microtime(true);

  // Display the results
  foreach ($duplicateIpCounts as $ipInfo) {
    $ip = $ipInfo['ip'];
    // Skip invalid IP
    if (!isValidIp($ip)) {
      continue;
    }
    // Check if execution time has exceeded the maximum allowed time
    $elapsedTime = microtime(true) - $startTime;
    if ($elapsedTime > $maxExecutionTime) {
      // Execution time exceeded
      echo "Execution time exceeded maximum allowed time of {$maxExecutionTime} seconds." . PHP_EOL;
      exit(0);
    }

    // Re-count the same IP
    $stmt = $pdo->prepare("SELECT COUNT(*) as count
                       FROM proxies
                       WHERE SUBSTR(proxy, 0, INSTR(proxy, ':')) = :ip
                       AND status != 'active'
                       AND status != 'untested'
                       AND last_check < datetime('now', '-7 days')");
    $stmt->bindParam(':ip', $ip, PDO::PARAM_STR);
    $stmt->execute();
    $count = $stmt->fetchColumn();

    if (intval($count) <= 1) {
      continue;
    }

    echo "Count of proxies with IP $ip: $count\n";
    // Fetch all rows matching the IP address (including port)
    // Exclude active proxies
    $stmt = $pdo->prepare("SELECT \"_rowid_\", * FROM \"main\".\"proxies\"
                       WHERE SUBSTR(proxy, 0, INSTR(proxy, ':')) = :ip
                       AND status != 'active'
                       AND status != 'untested'
                       AND last_check < datetime('now', '-7 days')
                       ORDER BY RANDOM() LIMIT 0, 49999;");
    $stmt->bindParam(':ip', $ip, PDO::PARAM_STR);
    $stmt->execute();
    $ipRows = $stmt->fetchAll(PDO::FETCH_ASSOC);

    if (!empty($ipRows)) {
      $keepRow = null;
      // shuffle ips
      shuffle($ipRows);
      foreach ($ipRows as $row) {
        $log = '';
        $proxy = $row['proxy'];
        $isAlreadyDead = false;
        if (!empty($row['last_check'])) {
          $deadStatus = $row['status'] == 'port-closed' || $row['status'] == 'dead';
          // treat already dead proxy more than 1 month
          $isAlreadyDead = $deadStatus && isDateRFC3339OlderThanHours($row['last_check'], 73000);
        }
        if ($row['status'] == 'active') {
          continue;
        }
        if (!isValidProxy(trim($proxy))) {
          // Proxy is invalid, delete the row
          $deleteStmt = $pdo->prepare("DELETE FROM proxies WHERE \"_rowid_\" = :id AND \"proxy\" = :proxy");
          $deleteStmt->bindParam(':id', $row['id'], PDO::PARAM_INT);
          $deleteStmt->bindParam(':proxy', $proxy, PDO::PARAM_STR);
          $deleteStmt->execute();

          $log = "[FILTER-PORT] $proxy invalid proxy [DELETED]\n";
        } else {
          if (!$keepRow) {
            // initialize keep row
            $keepRow = $row;
          }
          // Check if the proxy was already checked this month
          if (!wasCheckedThisMonth($pdo, $proxy)) {
            if (isPortOpen($proxy)) {
              echo "$proxy port open\n";
              $db->updateData($proxy, ['status' => 'untested'], false);
              // Update the last_check timestamp
              $lastCheck = date(DATE_RFC3339); // Assign date to a variable
              $updateStmt = $pdo->prepare("UPDATE proxies SET last_check = :last_check WHERE proxy = :proxy");
              $updateStmt->bindParam(':last_check', $lastCheck, PDO::PARAM_STR); // Use variable here
              $updateStmt->bindParam(':proxy', $proxy, PDO::PARAM_STR);
              $updateStmt->execute();
              // keep open port
              $keepRow = $row;
            } elseif ($keepRow['proxy'] !== $proxy) {
              if ($perform_delete || $isAlreadyDead) {
                // Delete closed port
                $deleteStmt = $pdo->prepare("DELETE FROM proxies WHERE id = :id AND proxy = :proxy");
                $deleteStmt->bindParam(':id', $row['id'], PDO::PARAM_INT);
                $deleteStmt->bindParam(':proxy', $proxy, PDO::PARAM_STR);
                $deleteStmt->execute();
                $log = "[FILTER-PORT] $proxy port closed [DELETED]" . PHP_EOL;
              } else {
                $db->updateData($proxy, ['status' => 'port-closed'], false);
                $log = "[FILTER-PORT] $proxy port closed" . PHP_EOL;
              }
            }
          } else {
            $log = "[FILTER-PORT] $proxy [SKIPPED]" . PHP_EOL;
          }
        }
        if (!empty($log)) {
          echo $log;
          if (count($argv) == 1) {
            append_content_with_lock(__DIR__ . '/proxyChecker.txt', $log);
          }
        }
      }
    }
  }

  // write working proxies
  write_working();

  // Increment start for the next batch
  $start += $batchSize;
} while (!empty($duplicateIpCounts));

// Close connection
$pdo = null;

// Function to check if the proxy was checked this month
function wasCheckedThisMonth(\PDO $pdo, string $proxy)
{
  $stmt = $pdo->prepare("SELECT COUNT(*) FROM proxies WHERE proxy = :proxy AND strftime('%Y-%m', last_check) = strftime('%Y-%m', 'now')");
  $stmt->bindParam(':proxy', $proxy, PDO::PARAM_STR);
  $stmt->execute();
  return $stmt->fetchColumn() > 0;
}

// Function to check if the proxy was checked this week
function wasCheckedThisWeek($pdo, $proxy)
{
  $startOfWeek = date('Y-m-d', strtotime('last sunday'));
  $stmt = $pdo->prepare("SELECT COUNT(*) FROM proxies WHERE proxy = :proxy AND last_check >= :start_of_week");
  $stmt->bindParam(':proxy', $proxy, PDO::PARAM_STR);
  $stmt->bindParam(':start_of_week', $startOfWeek, PDO::PARAM_STR);
  $stmt->execute();
  return $stmt->fetchColumn() > 0;
}

function write_working()
{
  global $db;
  echo "[FILTER-PORT] writing working proxies" . PHP_EOL;
  $data = parse_working_proxies($db);
  file_put_contents(__DIR__ . '/working.txt', $data['txt']);
  file_put_contents(__DIR__ . '/working.json', json_encode($data['array']));
  file_put_contents(__DIR__ . '/status.json', json_encode($data['counter']));
  return $data;
}

File: filterPortsDuplicate.py
import argparse
import random
import sqlite3
from datetime import datetime, timedelta
from sqlite3 import Cursor
from typing import Callable, Dict, List, Optional, Union
import concurrent.futures
from joblib import Parallel, delayed

from proxyCheckerReal import real_check
from src.func import get_relative_path
from src.func_console import green, red
from src.func_proxy import is_port_open
from src.ProxyDB import ProxyDB

# remove duplicate ip's more than 3 proxies

db = ProxyDB(get_relative_path("src/database.sqlite"), True)
conn = db.db.conn
cursor = conn.cursor()


def is_date_rfc3339_older_than_hours(date_str, hours):
    date = datetime.strptime(date_str, "%Y-%m-%dT%H:%M:%S%z")
    return datetime.now(date.tzinfo) - date > timedelta(hours=hours)


def was_checked_this_month(cursor: Cursor, proxy: str):
    cursor.execute(
        "SELECT COUNT(*) FROM proxies WHERE proxy = ? AND strftime('%Y-%m', last_check) = strftime('%Y-%m', 'now')",
        (proxy,),
    )
    return cursor.fetchone()[0] > 0


def was_checked_this_week(cursor: Cursor, proxy: str):
    start_of_week = (
        datetime.now() - timedelta(days=datetime.now().weekday())
    ).strftime("%Y-%m-%d")
    cursor.execute(
        "SELECT COUNT(*) FROM proxies WHERE proxy = ? AND last_check >= ?",
        (proxy, start_of_week),
    )
    return cursor.fetchone()[0] > 0


# Step 1: Identify and process duplicates based on IP address in batches
batch_size = 1000  # Adjust batch size as needed
start = 0
duplicate_ids = []
max_execution_time = 60  # Define maximum execution time
perform_delete = True


def fetch_proxies_same_ip(
    status: List[str] = ["dead", "port-closed", "untested"]
) -> Dict[str, List[str]]:
    """
    Fetch proxies that share the same IP but have different ports and match any of the given statuses.

    Args:
        status (List[str]): List of statuses to filter proxies. Defaults to ["dead", "port-closed", "untested"].

    Returns:
        Dict[str, List[str]]: A dictionary where keys are IP addresses and values are lists of proxies with different ports.
    """
    # Create a condition string from the status list
    condition = " OR ".join([f"status = '{s}'" for s in status])

    # Define the query to find proxies with the same IP but different ports
    query = f"""
    SELECT proxy
    FROM proxies
    WHERE SUBSTR(proxy, 1, INSTR(proxy, ':') - 1) IN (
        SELECT SUBSTR(proxy, 1, INSTR(proxy, ':') - 1)
        FROM proxies
        WHERE {condition}
        GROUP BY SUBSTR(proxy, 1, INSTR(proxy, ':') - 1)
        HAVING COUNT(*) > 1
    )
    ORDER BY SUBSTR(proxy, 1, INSTR(proxy, ':') - 1)
    """

    # Execute the query
    cursor.execute(query)

    # Fetch the results
    proxies: List[Dict[str, Union[str, None]]] = cursor.fetchall()

    # Process the results
    result: Dict[str, List[str]] = {}
    for row in proxies:
        proxy = row[0]
        ip = proxy.split(":")[0]
        if ip not in result:
            result[ip] = []
        result[ip].append(proxy)

    return result


def filter_duplicates_ips(max: int = 10, callback: Optional[Callable] = None):
    """
    Filter duplicated IPs by port open checks using multithreading.
    """
    duplicates_ips = fetch_proxies_same_ip()

    def process_ip(ip, ip_proxies):
        db = ProxyDB(get_relative_path("src/database.sqlite"), True)
        conn = db.db.conn
        cursor = conn.cursor()
        print(f"{ip} has {len(ip_proxies)} duplicates")

        if len(ip_proxies) > 1:
            # Fetch all rows matching the IP address (excluding active and port-open proxies)
            cursor.execute(
                """
                SELECT rowid, * FROM proxies
                WHERE SUBSTR(proxy, 1, INSTR(proxy, ':') - 1) = ?
                AND status != 'active' AND status != 'port-open'
                ORDER BY RANDOM() LIMIT 0, 49999;
                """,
                (ip,),
            )
            ip_rows = cursor.fetchall()

            if len(ip_rows) > 1:
                keep_row = ip_rows[0]
                random.shuffle(ip_rows)

                for row in ip_rows:
                    proxy = row["proxy"]
                    if is_port_open(proxy):
                        print(f"{proxy} {green('port open')}")
                        # keep open port
                        keep_row = row
                        # set status to port-open
                        last_check = datetime.now().strftime("%Y-%m-%dT%H:%M:%S%z")
                        cursor.execute(
                            "UPDATE proxies SET last_check = ?, status = ? WHERE proxy = ?",
                            (last_check, "port-open", proxy),
                        )
                        conn.commit()
                    else:
                        print(f"{proxy} {red('port closed')}")
                        if keep_row["proxy"] != proxy:
                            cursor.execute(
                                "DELETE FROM proxies WHERE proxy = ?", (proxy,)
                            )
                            conn.commit()

        db.close()

    with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
        futures = []
        for index, (ip, ip_proxies) in enumerate(duplicates_ips.items()):
            if index >= max:
                break
            futures.append(executor.submit(process_ip, ip, ip_proxies))

        for future in concurrent.futures.as_completed(futures):
            try:
                future.result()  # Ensure all threads complete
            except Exception as e:
                print(f"Error processing IP: {e}")

    if callable(callback):
        callback()


def worker_check_open_ports(item: Dict[str, str]):
    try:
        db = ProxyDB(get_relative_path("src/database.sqlite"), True)
        conn = db.db.conn
        cursor = conn.cursor()

        test = real_check(
            item["proxy"], "https://www.axis.co.id/bantuan", "pusat layanan"
        )
        if not test["result"]:
            test = real_check(item["proxy"], "https://www.example.com/", "example")

        if not test["result"]:
            test = real_check(item["proxy"], "http://azenv.net/", "AZ Environment")

        if not test["result"]:
            test = real_check(item["proxy"], "http://httpforever.com/", "HTTP Forever")

        last_check = datetime.now().strftime("%Y-%m-%dT%H:%M:%S%z")
        if test["result"]:
            cursor.execute(
                "UPDATE proxies SET last_check = ?, status = ? WHERE proxy = ?",
                (last_check, "active", item["proxy"]),
            )
            conn.commit()
        else:
            # cursor.execute("DELETE FROM proxies WHERE proxy = ?", (item["proxy"],))
            cursor.execute(
                "UPDATE proxies SET last_check = ?, status = ? WHERE proxy = ?",
                (last_check, "dead", item["proxy"]),
            )
            conn.commit()

    except sqlite3.Error as e:
        print(f"SQLite error occurred: {e}")
        # Handle the error as per your application's requirements
        # Example: Log the error, rollback transactions, etc.

    except Exception as e:
        print(f"Error occurred: {e}")
        # Handle other exceptions if needed

    finally:
        if "db" in locals():
            db.close()


def fetch_open_ports(max: int = 10) -> List[Dict[str, Union[str, None]]]:
    global cursor
    # Execute your query using the global cursor
    cursor.execute("SELECT proxy FROM proxies WHERE status = 'port-open'")
    proxies = cursor.fetchall()

    # Fetch column names
    column_names = [description[0] for description in cursor.description]

    # Convert the result into a list of dictionaries
    proxies_dict: List[Dict[str, Union[str, None]]] = [
        dict(zip(column_names, proxy)) for proxy in proxies
    ][:max]

    return proxies_dict


def check_open_ports(max: int = 10, callback: Optional[Callable] = None):
    """
    check proxy WHERE status = 'port-open'
    """
    proxies_dict = fetch_open_ports(max)
    if len(proxies_dict) > 0:
        Parallel(n_jobs=10)(
            delayed(worker_check_open_ports)(item) for item in proxies_dict
        )
    if callable(callback):
        callback()


def remove_duplicates_dead_proxies(max: int = 10, callback: Optional[Callable] = None):
    duplicates_ips: Dict[str, List[str]] = fetch_proxies_same_ip(
        ["dead", "port-closed"]
    )

    def process_ip(ip: str, ip_proxies: List[str]):
        print(f"Processing ip: {ip}")
        db = ProxyDB(get_relative_path("src/database.sqlite"), True)
        # keep first row (dont delete first row)
        ip_proxies_excluding_first = ip_proxies[1:]
        for proxy in ip_proxies_excluding_first:
            db.remove(proxy)
            print(f"{proxy} {red('removed')}")
        db.close()

    with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
        futures = []
        for index, (ip, ip_proxies) in enumerate(duplicates_ips.items()):
            if index >= max:
                break
            futures.append(executor.submit(process_ip, ip, ip_proxies))

        for future in concurrent.futures.as_completed(futures):
            try:
                future.result()  # Ensure all threads complete
            except Exception as e:
                print(f"Error processing IP: {e}")
    if callable(callback):
        callback()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Proxy Tool")
    parser.add_argument("--max", type=int, help="Maximum number of proxies to check")
    args = parser.parse_args()
    max = 1
    if args.max:
        max = args.max

    filter_duplicates_ips(
        max, lambda: check_open_ports(max, lambda: remove_duplicates_dead_proxies(max))
    )

    # Close connection
    db.close()
